shared/src/main/scala/cache.scala

```diff
@@ -28,42 +30,49 @@ final class DataSourceResult(val result: Any) extends AnyVal
 /**
  * A `Cache` trait so the users of the library can provide their own cache.
  */
-trait DataSourceCache {
-  def lookup[F[_] : ConcurrentEffect, I, A](i: I, ds: DataSource[I, A]): F[Option[A]]
-  def insert[F[_] : ConcurrentEffect, I, A](i: I, v: A, ds: DataSource[I, A]): F[DataSourceCache]
-  def insertMany[F[_]: ConcurrentEffect, I, A](vs: Map[I, A], ds: DataSource[I, A]): F[DataSourceCache] =
-    vs.toList.foldLeftM(this)({
-      case (c, (i, v)) => c.insert(i, v, ds)
-    })
+trait DataSourceCache[F[_]] {
+  def lookup[I, A](i: I, ds: DataSource[I, A])(
+    implicit C: ConcurrentEffect[F], P: Par[F]
```

Comment:
It would be nice if we didn't put these constraints on the methods anymore.

Now that we have parametrized `DataSourceCache` on `F`, the implementation could specify the constraints.
For example for `InMemoryCache` we would only need `Applicative` (`Monad` for `foldLeftM`).

Reply:
Good point, I'm going to try to remove these constraints.